File: remove_nth/remove_nth.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        if not head or not head.next: return None
        
        # fast is the lead element. remove is the element to remove
        # prev is the element before remove
        fast = head
        remove = head
        prev = None
        
        for i in range(n - 1): fast = fast.next
        
        # we must be ditching the first element
        if not fast.next: return head.next
        
        # otherwise, create prev -> remove -> ... -> fast, then point
        # prev.next to remove.next (cutting out remove)
        while fast.next:
            prev = remove
            remove = remove.next
            fast = fast.next    
        prev.next = remove.next
        return head

File: remove_nth/RemoveNth.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        
        // One pass algorithm: move one pointer n nodes ahead.
        // Then, until the `front` pointer reaches the end,
        // advance both pointers by one. 
        // Then, the `back` pointer will be pointing
        // at the node one before the node to be removed.
        ListNode front = head;
        ListNode back  = head;
        
        for (int i = 0; i < n; i++) {
            front = front.next;
        }
        if (front == null)  {
            head = head.next;
            return head;
        }
        
        while (front.next != null) {
            front = front.next;
            back = back.next;
        }
        
        back.next = back.next.next;
        return head;
    }
}